apps/prairielearn/assets/scripts/lib/confirmOnUnload.ts

```diff
@@ -0,0 +1,37 @@
+function getQuestionFormData(form: HTMLFormElement): string {
+  // Cast FormData since TS does not support this parameter,
+  // see https://github.com/microsoft/TypeScript/issues/30584
+  const formData = new URLSearchParams(new FormData(form) as any);
+  formData.delete('__csrf_token');
+  form.querySelectorAll('[data-skip-unload-check]').forEach((input: HTMLInputElement) => {
+    if (input.name) formData.delete(input.name);
+  });
+  return formData.toString();
+}
+
+export function saveQuestionFormData(form: HTMLFormElement) {
+  if (!form) return;
+  form.dataset.originalFormData = getQuestionFormData(form);
+}
+
+export function confirmOnUnload(form: HTMLFormElement) {
+  // Set form state on load and submit
+  saveQuestionFormData(form);
+  form.addEventListener('submit', () => {
+    saveQuestionFormData(form);
+  });
+
+  // Check form state on unload
+  window.addEventListener('beforeunload', (event) => {
+    const isSameForm = form.dataset.originalFormData === getQuestionFormData(form);
+
+    if (!isSameForm) {
+      // Supported in most modern browsers
+      event.preventDefault();
+
+      // Fallback for legacy browsers
+      event.returnValue = '';
+      return '';
```

Comment:
We are wayyyyyy past supporting any legacy browser that doesn't support `preventDefault()` given that we're using modern features like `fetch`. Let's drop this.

Reply:
According to https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event and https://caniuse.com/mdn-api_window_beforeunload_event_preventdefault_activation, most modern browsers don't support preventDefault in this case either, including Chrome. I'll test further later, I'm on mobile now, and I'll update the comments accordingly.

Reply:
Fixed comments in 36a4224f52a3a60316985982cc9a37df86a8b4a9. It is still necessary to return a string or set `returnValue`, but not both, so I opted for returning a string.